storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java

```diff
@@ -0,0 +1,1115 @@
+                                    newOffset, expectedNextOffset, logEndOffset(), activeSegment.timeIndex().entries(), activeSegment.offsetIndex().entries());
+                            LogSegment newSegment = createAndDeleteSegment(newOffset, activeSegment, true, new LogRoll(this.logger));
+                            updateLogEndOffset(nextOffsetMetadata.messageOffset);
+                            logger.info("Rolled new log segment at offset {} in {} ms.", newOffset, time.hiResClockMs() - start);
+                            return newSegment;
+                        } else {
+                            throw new KafkaException("Trying to roll a new log segment for topic partition "
+                                    + topicPartition + " with start offset " + newOffset + " + =max(provided offset = "
+                                    + expectedNextOffset + ", LEO = " + logEndOffset() + " while it already exists. Existing segment is "
+                                    + segments.get(newOffset) + ".");
+                        }
+                    } else if (!segments.isEmpty() && newOffset < activeSegment.baseOffset()) {
+                        throw new KafkaException("Trying to roll a new log segment for topic partition " + topicPartition
+                                + " with start offset " + newOffset + " =max(provided offset = " + expectedNextOffset
+                                + ", LEO = " + logEndOffset() + ") lower than start offset of the active segment " + activeSegment);
+                    } else {
+                        File offsetIdxFile = LogFileUtils.offsetIndexFile(dir, newOffset);
+                        File timeIdxFile = LogFileUtils.timeIndexFile(dir, newOffset);
+                        File txnIdxFile = LogFileUtils.transactionIndexFile(dir, newOffset);
+
+                        for (File file : Arrays.asList(logFile, offsetIdxFile, timeIdxFile, txnIdxFile)) {
+                            if (file.exists()) {
+                                logger.warn("Newly rolled segment file {} already exists; deleting it first", file.getAbsolutePath());
+                                Files.delete(file.toPath());
+                            }
+                        }
+
+                        Optional<LogSegment> logSegmentOption = segments.lastSegment();
+                        if (logSegmentOption.isPresent()) logSegmentOption.get().onBecomeInactiveSegment();
+                    }
+
+                    LogSegment newSegment = LogSegment.open(dir,
+                            newOffset,
+                            config,
+                            time,
+                            false,
+                            config.initFileSize(),
+                            config.preallocate,
+                            "");
```

Comment:
Why are we using a different overload of `open` here?

Reply:
Good catch! We can reuse the existing `open` method.